sparkplug-core/src/java/sparkplug/function/SerializableFn.java

```diff
@@ -5,56 +5,146 @@
 import clojure.lang.IFn;
 import clojure.lang.RT;
 import clojure.lang.Symbol;
+import clojure.lang.Var;
 
 import java.io.IOException;
 import java.io.InvalidObjectException;
+
 import java.io.ObjectInputStream;
 import java.io.ObjectOutputStream;
 import java.io.Serializable;
 
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 
 /**
  * Base class for function classes built for interop with Spark and Scala.
+ *
+ * This class is designed to be serialized across computation boundaries in a
+ * manner compatible with Spark and Kryo, while ensuring that required code is
+ * loaded upon deserialization.
  */
 public abstract class SerializableFn implements Serializable {
 
+    private static final Logger logger = LoggerFactory.getLogger(SerializableFn.class);
+    private static final Var require = RT.var("clojure.core", "require");
+
     protected IFn f;
+    protected List<String> namespaces;
 
 
     /**
      * Default empty constructor.
+     *
+     * TODO is this accurate?
+     * Inheriting classes must also have an empty constructor to be properly
+     * serializable.
      */
     private SerializableFn() {
     }
 
 
-    private void writeObject(ObjectOutputStream out) throws IOException {
+    /**
+     * Construct a new serializable wrapper for the function with an explicit
+     * set of required namespaces.
+     *
+     * @param fn Clojure function to wrap
+     * @param namespaces collection of namespaces required
+     */
+    protected SerializableFn(IFn fn, Collection<String> namespaces) {
+        this.f = fn;
+        this.namespaces = new ArrayList<String>(namespaces);
         // Attempt to derive the needed Clojure namespace from the function's class name.
-        String namespace = Compiler.demunge(this.f.getClass().getName()).split("/")[0];
-        out.writeObject(namespace);
-        out.writeObject(this.f);
+        String fnNamespace = Compiler.demunge(fn.getClass().getName()).split("/")[0];
+        this.namespaces.add(fnNamespace);
```

Comment:
This might result in `fnNamespace` existing twice in the list; you could use a set here or (probably better) move the ns-guessing code into clojure with the rest of the field-walking logic. I think this also adds `clojure.core` back to the namespaces if `f` is a higher-order core function.